plugins/legion-hid2/fu-legion-hid2-device.c

```diff
@@ -0,0 +1,522 @@
+}
+
+static gboolean
+fu_legion_hid2_device_ensure_version(FuDevice *device, GError **error)
+{
+	guint32 version;
+	g_autoptr(GByteArray) cmd = fu_struct_legion_get_version_new();
+	g_autoptr(GByteArray) result = fu_struct_legion_version_new();
+
+	if (!fu_legion_hid2_device_transfer(FU_LEGION_HID2_DEVICE(device), cmd, result, error))
+		return FALSE;
+	version = fu_struct_legion_version_get_version(result);
+	fu_device_set_version_raw(device, version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_ensure_mcu_id(FuDevice *device, GError **error)
+{
+	g_autoptr(GByteArray) cmd = fu_struct_legion_get_mcu_id_new();
+	g_autoptr(GByteArray) result = fu_struct_legion_mcu_id_new();
+
+	if (!fu_legion_hid2_device_transfer(FU_LEGION_HID2_DEVICE(device), cmd, result, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_probe(FuDevice *device, GError **error)
+{
+	if (fu_device_has_flag(FU_DEVICE(device), FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		fu_hid_device_set_interface(FU_HID_DEVICE(device), 0);
+		fu_hid_device_set_ep_addr_in(FU_HID_DEVICE(device), 0x81);
+		fu_hid_device_set_ep_addr_out(FU_HID_DEVICE(device), 1);
+	} else {
+		fu_hid_device_set_interface(FU_HID_DEVICE(device), 3);
+		fu_hid_device_set_ep_addr_in(FU_HID_DEVICE(device), 0x84);
+		fu_hid_device_set_ep_addr_out(FU_HID_DEVICE(device), 0x4);
+	}
+
+	/* FuHidDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_legion_hid2_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_setup(FuDevice *device, GError **error)
+{
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_legion_hid2_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* can't use anything but write and reset in IAP mode */
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER))
+		return TRUE;
+
+	if (!fu_legion_hid2_device_ensure_version(device, error))
+		return FALSE;
+
+	if (!fu_legion_hid2_device_ensure_mcu_id(device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_legion_hid2_device_prepare_firmware(FuDevice *device,
+				       GInputStream *stream,
+				       FuProgress *progress,
+				       FwupdInstallFlags flags,
+				       GError **error)
+{
+	guint32 version;
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(FuFirmware) firmware = fu_legion_hid2_firmware_new();
+
+	/* sanity check */
+	if (!fu_firmware_parse_stream(firmware, stream, 0x0, flags, error))
+		return NULL;
+
+	version = fu_legion_hid2_firmware_get_version(firmware);
+	if (fu_device_get_version_raw(device) > version) {
+		version_str = fu_version_from_uint32(version, FWUPD_VERSION_FORMAT_QUAD);
+		if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "downgrading from %s to %s is not supported",
+				    fu_device_get_version(device),
+				    version_str);
+			return NULL;
+		}
```

Comment:
Do you want to set FWUPD_DEVICE_FLAG_ONLY_VERSION_UPGRADE as well?

Reply:
No; because reinstall is totally fine.